src/cpp/r/include/r/RCntxt.hpp

```diff
@@ -0,0 +1,128 @@
+/*
+ * RCntxt.hpp
+ *
+ * Copyright (C) 2009-16 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#ifndef R_CONTEXT_HPP
+#define R_CONTEXT_HPP
+
+#include "RSexp.hpp"
+#include "RCntxtInterface.hpp"
+
+#include <core/Error.hpp>
+
+#include <boost/iterator/iterator_facade.hpp>
+
+namespace rstudio {
+namespace r {
+namespace context {
+
+// RCnxt is a public-facing class which wraps entries from the R context
+// stack (RCNXT* in the R headers), which keeps track of execution contexts
+// during the evaluation of R code. Unfortunately, the memory layout
+// in the RCNXT* struct is not identical in all R versions, and consequently
+// it's necessary to wrap them in a layer of abstraction, which this class
+// supplies.
+//
+// It has value semantics; its only data member is a shared pointer to the class
+// managing the RNCTXT* interface. It also implements simple iteration over
+// entries in the context stack. RCnxt::begin() returns an iterator beginning
+// at R_GlobalContext; RCnxt::end() returns a null context entry.
+class RCntxt: public RCntxtInterface
+{
+public:
+   // copy/equality testing
+   RCntxt();
+   RCntxt(void *rawCntxt);
+   bool operator==(const RCntxt& other) const;
+   operator bool() const { return pCntxt_; }
```

Comment:
FWIW the reason we do that is to avoid unintentionally casting to non-bool primitive types, e.g. right now this would compile:

```
int ctx = pCntxt_;
```

Also FWIW C++11 will allow us to write

```
explicit operator bool() const { return pCntxt_; }
```